### PR TITLE
Pserve --reload does not keep worker arguments. Fixes #2961

### LINKED ISSUE #2961: pserve --reload call hupper which call pserve with sys.argv instead of the parameters
If you are calling `pserve.main(argv)` and `argv` contains `--reload` it will call upper with `sys.argv` instead of `argv`.

### PR DESCRIPTION
**Pull request by Natim**
Refs #2961

### CODE REVIEW

**Comment by bertjwregeer** [comment]
Please don't mix whitespace/pep8 fixes in with the rest of the commit in the future. It makes it more difficult to see what the changes are.

File: pyramid/tests/test_scripts/test_pserve.py [hunk 1]
```diff
@@ -1,9 +1,11 @@
+import mock
```
Comment (bertjwregeer) [inline pyramid/tests/test_scripts/test_pserve.py]: We don't currently use mock in the the Pyramid tests, instead we create dummy objects that have just the functionality required.

I don't think it is a great idea to add yet another dependency...

File: setup.py [hunk 2]
```diff
@@ -65,9 +65,10 @@
     ]
 
 testing_extras = tests_require + [
+    'mock',
```
Comment (bertjwregeer) [inline setup.py]: I'm a -1 on this, we use dummy objects everywhere else.

**Comment by Natim** [comment]
> Please don't mix whitespace/pep8 fixes in with the rest of the commit in the future. It makes it more difficult to see what the changes are.

That's why I put them in another commit so that you can look at the diff without them.

> I don't think it is a great idea to add yet another dependency...

It doesn't change anything to add another dependency, but I will try to rewrite my test with a Dummy object instead. (mock just makes it easy to do so with less code to maintain.)

File: setup.py (hunk 2 above)
Comment (Natim) [inline setup.py]: By the way this dependencies are only installed when you run tests which is not something you do when you use the project.

Comment (bertjwregeer) [inline setup.py]: Sure, I understand that.

**Comment by mmerickel** [comment]
ping @Natim

**Comment by Natim** [comment]
Hi @mmerickel, I didn't forget about this and I am going to rewrite tests without mocks.

**Comment by mmerickel** [comment]
@Natim Okie sounds good.. just wanted to make sure it wasn't forgotten!

**Comment by Natim** [comment]
To be honest I find the code much harder to maintain with a test like that. Are you sure it worth getting rid of the mock dependency?

**Comment by Natim** [comment]
@bertjwregeer @mmerickel how does it look to you now?

**Comment by Natim** [comment]
Any news about it?

File: pyramid/tests/test_scripts/test_pserve.py [hunk 3]
```diff
@@ -77,6 +80,32 @@ def test_config_file_finds_watch_files(self):
...
+
+        def dummy_start_reloader(*args, **kwargs):
+            dummy_start_reloader.args = args
+            dummy_start_reloader.kwargs = kwargs
+
+        pserve.hupper = AttrDict(is_active=lambda: False,
+                                 start_reloader=dummy_start_reloader)
+
+        inst = self._makeOne('--reload', 'development.ini')
+        inst.run()
```
Comment (mmerickel) [inline pyramid/tests/test_scripts/test_pserve.py]: I know we aren't using mock (sorry!) but if you're patching a module-scope value, you need to replace it after the test is over. For example:

```
orig_hupper = pserve.hupper
try:
    pserve.hupper = ...
    inst = ...
    inst.run()
finally:
    pserve.hupper = orig_hupper
```

**Comment by Natim** [comment]
> I know we aren't using mock (sorry!) 

Why cannot we start?

**Comment by mmerickel** [comment]
> Why cannot we start?

Well I'm not adding a new dependency in a bugfix release and I imagine we want to backport this to 1.8. I'm definitely open to improving this situation in 1.9 though.

**Comment by Natim** [comment]
> Well I'm not adding a new dependency in a bugfix release and I imagine we want to backport this to 1.8. I'm definitely open to improving this situation in 1.9 though.

Ok it makes sense 👍

**Comment by mmerickel** [comment]
Just waiting on @bertjwregeer to approve this before I continue.

@Natim if you come back with more bugs the day after I cut a new release again I'm going to bonk you on the head if I ever see you in person!

**Comment by bertjwregeer** [comment]
I will say that I am not all that excited about using Mock. I feel like mock is overused for "magic" to make tests pass, and it just overall makes a bad experience for writing tests. I think we can try to use it, but we have to be careful to not be overly reliant on it.

**Comment by Natim** [comment]
Hopefully releases are cheap :)

**Comment by mmerickel** [comment]
@Natim I wish that were true. :-(

**Pull request merged**